spring/src/main/java/io/seata/spring/annotation/datasource/SeataAutoDataSourceProxyCreator.java

```diff
@@ -0,0 +1,84 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.spring.annotation.datasource;
+
+import javax.sql.DataSource;
+import java.util.Arrays;
+
+import io.seata.rm.datasource.DataSourceProxy;
+import io.seata.spring.util.SpringProxyUtils;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.aop.Advisor;
+import org.springframework.aop.TargetSource;
+import org.springframework.aop.framework.AdvisedSupport;
+import org.springframework.aop.framework.autoproxy.AbstractAutoProxyCreator;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.aop.support.DefaultIntroductionAdvisor;
+import org.springframework.beans.BeansException;
+
+/**
+ * @author xingfudeshi@gmail.com
+ */
+public class SeataAutoDataSourceProxyCreator extends AbstractAutoProxyCreator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SeataAutoDataSourceProxyCreator.class);
+    private MethodInterceptor advice;
+    private final String[] excludes;
+
+    public SeataAutoDataSourceProxyCreator(boolean useJdkProxy, String[] excludes) {
+        this.excludes = excludes;
+        setProxyTargetClass(!useJdkProxy);
+    }
+
+    @Override
+    protected Object wrapIfNecessary(Object bean, String beanName, Object cacheKey) {
```

Comment:
What I thought is to not override `wrapIfNecessary` and fully make use of `AbstractAutoProxyCreator`, something like this:

```
public class SeataAutoDataSourceProxyCreator extends AbstractAutoProxyCreator {
    private final String[] excludes;

    private final DefaultIntroductionAdvisor advisor = new DefaultIntroductionAdvisor(new SeataAutoDataSourceProxyAdvice());

    public SeataAutoDataSourceProxyCreator(boolean useJdkProxy, String[] excludes) {
        this.excludes = excludes;
        setProxyTargetClass(!useJdkProxy);
    }

    @Override
    protected Object[] getAdvicesAndAdvisorsForBean(Class<?> beanClass, String beanName, TargetSource customTargetSource) throws BeansException {
        return new Object[]{advisor};
    }

    @Override
    protected boolean shouldSkip(Class<?> beanClass, String beanName) {
        return SeataProxy.class.isAssignableFrom(beanClass) ||
                !DataSource.class.isAssignableFrom(beanClass) ||
                Arrays.asList(excludes).contains(beanClass.getName());
    }
}

public class SeataAutoDataSourceProxyAdvice implements MethodInterceptor, IntroductionInfo {
    @Override
    public Object invoke(MethodInvocation invocation) throws Throwable {
        DataSource dataSource = (DataSource) invocation.getThis();
        DataSourceProxy dataSourceProxy = DataSourceProxyHolder.get().putDataSource(dataSource);
        Method method = invocation.getMethod();
        Object[] args = invocation.getArguments();
        Method m = BeanUtils.findDeclaredMethod(DataSourceProxy.class, method.getName(), method.getParameterTypes());
        if (null != m) {
            return m.invoke(dataSourceProxy, args);
        } else {
            return invocation.proceed();
        }
    }

    @Override
    public Class<?>[] getInterfaces() {
        return new Class[]{SeataProxy.class};
    }

}
```

Reply:
Fixed.Thanks.